### ISSUE TITLE
URI not working with Google Authenticator, as per demo

### ISSUE THREAD

**Issue by DavidFricker** [Spomky-Labs/otphp#104]
| Q                    | A
| -------------------- | -----
| Bug report?          | yes
| Library version      | 9.1.0

The rendered QR code from this call `$totp->getQrCodeUri()` is not accepted by Google Authenticator.
Error message from the app: "the barcode 'optauth://totp/ ... " is not a valid authentication token barcode"

Some investigation leads me to believe the URI contained in the QR is not of correct format as per [https://github.com/google/google-authenticator/wiki/Key-Uri-Format](https://github.com/google/google-authenticator/wiki/Key-Uri-Format)

- The secret is not base32 encoded

**Comment by Spomky** [comment]
Hi,

Can you elaborate a bit more by providing an example?

I just wrote the following lines and it works great on Google Authenticator (v5.00, Android) and the version 9.1.0 of the library.

```php
use OTPHP\TOTP;

$totp = TOTP::create();
$totp->setLabel('email@example.com');
$totp->setIssuer('Example');

$googleChartUri = $totp->getQrCodeUri();
echo sprintf('<img src="%s">', $googleChartUri);
```

**Comment by DavidFricker** [comment]
Thanks that solved my issue! Knew I was doing something stupidly wrong.

The root cause was found to be that I specified the secret manually, without knowing that the secret should be Base32 encoded when passed to the `create` method.

Sorry to bother you

**Comment by Spomky** [comment]
OK so closing as resolved.
The need for a Base32 encoded secret is mentionned in https://github.com/Spomky-Labs/otphp/blob/v9.0/doc/Customize.md#secret

**Comment by webmaster777** [comment]
If the secret contains padding chars (=), this will also create a problem on at least the iOS Google Authenticator app. I've managed to work around it by using `Base32::encodeUpperUnpadded($key)` as secret, but I think the uri-generator should not allow padding characters to apear in the key (as suggested by https://github.com/google/google-authenticator/wiki/Key-Uri-Format#secret)

**Comment by apm963** [comment]
@webmaster777 One of my users just experienced this issue on iOS. Your solution of using `Base32::encodeUpperUnpadded` (by ignoring RFC 3548 section 2.2 as per provided link) was exactly what fixed this for us. I am using version 8.3.x, though, so I'm not sure if a new issue needs to be opened or if this has been addressed in newer versions.

**Comment by webmaster777** [comment]
We're using the latest version, and it is not adressed.

**Comment by Spomky** [comment]
Hi,

You are right to say that this is not clearly mentioned in the documentation.
I will add some words about that issue.

If you don’t set a custom secret (i.e. the library create it), [the trailing `=` are always removed](https://github.com/Spomky-Labs/otphp/blob/v9.0/src/ParameterTrait.php#L208) on both v8.x and v9.x.

The library don’t use `Base32::encodeUpperUnpadded` because this method was introduced in [paragonie/constant_time_encoding v2.2.0](https://github.com/paragonie/constant_time_encoding/releases/tag/v2.2.0) whilst the library may depends on lower versions (requires at least v2.0.0).

**Comment by Spomky** [comment]
The problem still exist when provisioning uri are loaded. I will also remove these characters automatically.

**Comment by Spomky** [comment]
I’ve just released `v9.1.1` that should fix this.
